File: algorithms/easy/125_valid_palindrome.py
# 125. Valid Palindrome

def is_palindrome(s):
    """Determine if a string is a palindrome, considering only alphanumeric characters.
    
    A palindrome reads the same backward as forward.
    
    Args:
        s: String to check
    
    Returns:
        Boolean indicating if the string is a palindrome
    """
    # Convert to lowercase and filter out non-alphanumeric characters
    filtered_chars = [c.lower() for c in s if c.isalnum()]
    
    # Check if the filtered string is equal to its reverse
    return filtered_chars == filtered_chars[::-1]

# Alternative implementation using two pointers
def is_palindrome_two_pointers(s):
    # Convert to lowercase and filter out non-alphanumeric characters
    filtered_chars = [c.lower() for c in s if c.isalnum()]
    
    # Use two pointers to compare characters from both ends
    left, right = 0, len(filtered_chars) - 1
    
    while left < right:
        if filtered_chars[left] != filtered_chars[right]:
            return False
        left += 1
        right -= 1
    
    return True

# More efficient implementation without creating a new list
def is_palindrome_efficient(s):
    left, right = 0, len(s) - 1
    
    while left < right:
        # Skip non-alphanumeric characters from the left
        while left < right and not s[left].isalnum():
            left += 1
        
        # Skip non-alphanumeric characters from the right
        while left < right and not s[right].isalnum():
            right -= 1
        
        # Compare characters (case-insensitive)
        if s[left].lower() != s[right].lower():
            return False
        
        left += 1
        right -= 1
    
    return True

# Test cases
print(is_palindrome("A man, a plan, a canal: Panama"))  # True
print(is_palindrome("race a car"))                     # False
print(is_palindrome(" "))                              # True

print(is_palindrome_two_pointers("A man, a plan, a canal: Panama"))  # True
print(is_palindrome_efficient("A man, a plan, a canal: Panama"))     # True

File: algorithms/easy/125_valid_palindrome.js
/**
 * 125. Valid Palindrome
 * 
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    // Convert to lowercase and filter out non-alphanumeric characters
    const filteredChars = s.toLowerCase().replace(/[^a-z0-9]/g, '');
    
    // Check if the filtered string is equal to its reverse
    return filteredChars === filteredChars.split('').reverse().join('');
};

// Alternative implementation using two pointers
var isPalindromeTwoPointers = function(s) {
    // Convert to lowercase and filter out non-alphanumeric characters
    const filteredChars = s.toLowerCase().replace(/[^a-z0-9]/g, '');
    
    // Use two pointers to compare characters from both ends
    let left = 0;
    let right = filteredChars.length - 1;
    
    while (left < right) {
        if (filteredChars[left] !== filteredChars[right]) {
            return false;
        }
        left++;
        right--;
    }
    
    return true;
};

// More efficient implementation without creating a new string
var isPalindromeEfficient = function(s) {
    let left = 0;
    let right = s.length - 1;
    
    while (left < right) {
        // Skip non-alphanumeric characters from the left
        while (left < right && !isAlphanumeric(s[left])) {
            left++;
        }
        
        // Skip non-alphanumeric characters from the right
        while (left < right && !isAlphanumeric(s[right])) {
            right--;
        }
        
        // Compare characters (case-insensitive)
        if (s[left].toLowerCase() !== s[right].toLowerCase()) {
            return false;
        }
        
        left++;
        right--;
    }
    
    return true;
};

// Helper function to check if a character is alphanumeric
function isAlphanumeric(char) {
    const code = char.charCodeAt(0);
    return (code >= 48 && code <= 57) ||  // 0-9
           (code >= 65 && code <= 90) ||  // A-Z
           (code >= 97 && code <= 122);   // a-z
}

// Test cases
console.log(isPalindrome("A man, a plan, a canal: Panama"));  // true
console.log(isPalindrome("race a car"));                     // false
console.log(isPalindrome(" "));                              // true

console.log(isPalindromeTwoPointers("A man, a plan, a canal: Panama"));  // true
console.log(isPalindromeEfficient("A man, a plan, a canal: Panama"));     // true